File: PokePlannerWeb/ClientApp/src/models/EncountersEntry.ts
import { EncounterConditionValueEntry } from "./EncounterConditionValueEntry"
import { EncounterMethodEntry } from "./EncounterMethodEntry"
import { LocalString } from "./LocalString"
import { WithId } from "./WithId"

/**
 * Represents a Pokemon's encounters in the data store.
 */
export interface EncountersEntry {
    /**
     * The ID of the Pokemon.
     */
    pokemonId: number

    /**
     * The encounters indexed by version group ID.
     */
    encounters: WithId<EncounterEntry[]>[]
}

/**
 * Represents an encounter in the data store.
 */
export class EncounterEntry {
    /**
     * The ID of the location area of the encounter.
     */
    locationAreaId: number

    /**
     * The display names of the encounter.
     */
    displayNames: LocalString[]

    /**
     * The details of the encounter indexed by version ID.
     */
    details: WithId<EncounterMethodDetails[]>[]

    /**
     * Constructor.
     */
    constructor(
        locationAreaId: number,
        displayNames: LocalString[],
        details: WithId<EncounterMethodDetails[]>[]
    ) {
        this.locationAreaId = locationAreaId
        this.displayNames = displayNames
        this.details = details
    }

    /**
     * Returns a encounter entry created from the given entry.
     */
    static from(encounter: EncounterEntry) {
        return new EncounterEntry(
            encounter.locationAreaId,
            encounter.displayNames,
            encounter.details.map(e => new WithId<EncounterMethodDetails[]>(
                e.id,
                e.data.map(EncounterMethodDetails.from)
            ))
        )
    }

    /**
     * Returns the encounter's display name in the given locale.
     */
    getDisplayName(locale: string): string | undefined {
        let localName = this.displayNames.find(n => n.language === locale)
        if (localName === undefined) {
            console.warn(`Encounter is missing display name in locale '${locale}'`)
        }

        return localName?.value
    }
}

/**
 * Represents details of an encounter method.
 */
export class EncounterMethodDetails {
    /**
     * The encounter method.
     */
    method: EncounterMethodEntry

    /**
     * The encounter details.
     */
    conditionValuesDetails: ConditionValuesDetail[]

    /**
     * Constructor.
     */
    constructor(
        method: EncounterMethodEntry,
        conditionValuesDetails: ConditionValuesDetail[]
    ) {
        this.method = method
        this.conditionValuesDetails = conditionValuesDetails
    }

    /**
     * Returns a encounter method details object created from the given object.
     */
    static from(encounter: EncounterMethodDetails) {
        return new EncounterMethodDetails(
            EncounterMethodEntry.from(encounter.method),
            encounter.conditionValuesDetails.map(ConditionValuesDetail.from)
        )
    }
}

/**
 * Model for a list of encounter details under a given set of conditions.
 */
export class ConditionValuesDetail {
    /**
     * The condition values required for the encounters to occur.
     */
    conditionValues: EncounterConditionValueEntry[]

    /**
     * The encounter details.
     */
    encounterDetails: EncounterDetailEntry[]

    /**
     * Constructor.
     */
    constructor(
        conditionValues: EncounterConditionValueEntry[],
        encounterDetails: EncounterDetailEntry[]
    ) {
        this.conditionValues = conditionValues
        this.encounterDetails = encounterDetails
    }

    /**
     * Returns an encounter detail created from the given entry.
     */
    static from(detail: ConditionValuesDetail) {
        return new ConditionValuesDetail(
            detail.conditionValues.map(EncounterConditionValueEntry.from),
            detail.encounterDetails.map(EncounterDetailEntry.from)
        )
    }
}

/**
 * Represents an encounter.
 */
export class EncounterDetailEntry {
    /**
     * The chance of the encounter.
     */
    chance: number

    /**
     * The minimum level of the encounter.
     */
    minLevel: number

    /**
     * The maximum level of the encounter.
     */
    maxLevel: number

    /**
     * Constructor.
     */
    constructor(
        chance: number,
        minLevel: number,
        maxLevel: number
    ) {
        this.chance = chance
        this.minLevel = minLevel
        this.maxLevel = maxLevel
    }

    /**
     * Returns an encounter detail created from the given entry.
     */
    static from(detail: EncounterDetailEntry) {
        return new EncounterDetailEntry(
            detail.chance,
            detail.minLevel,
            detail.maxLevel
        )
    }
}

File: PokePlannerWeb.Data/DataStore/Models/EncountersEntry.cs
using System.Collections.Generic;
using PokeApiNet;

namespace PokePlannerWeb.Data.DataStore.Models
{
    /// <summary>
    /// Represents a Pokemon's encounters in the data store.
    /// </summary>
    public class EncountersEntry : EntryBase
    {
        /// <summary>
        /// Gets the ID of the Pokemon.
        /// </summary>
        public int PokemonId => Key;

        /// <summary>
        /// Gets or sets the encounters indexed by version group ID.
        /// </summary>
        public List<WithId<EncounterEntry[]>> Encounters { get; set; }
    }

    /// <summary>
    /// Represents an encounter in the data store.
    /// </summary>
    public class EncounterEntry
    {
        /// <summary>
        /// Gets or sets the ID of the location area of the encounter.
        /// </summary>
        public int LocationAreaId { get; set; }

        /// <summary>
        /// Gets or sets the display names of the encounter.
        /// </summary>
        public List<LocalString> DisplayNames { get; set; }

        /// <summary>
        /// Gets or sets the encounter chances indexed by version ID.
        /// </summary>
        public List<WithId<int>> Chances { get; set; }

        /// <summary>
        /// Gets or sets the details of the encounter indexed by version ID.
        /// </summary>
        public List<WithId<EncounterMethodDetails[]>> Details { get; set; }
    }

    /// <summary>
    /// Represents details of an encounter method.
    /// </summary>
    public class EncounterMethodDetails
    {
        /// <summary>
        /// Gets or sets the encounter method.
        /// </summary>
        public EncounterMethodEntry Method { get; set; }

        /// <summary>
        /// Gets or sets the encounter details.
        /// </summary>
        public List<Encounter> EncounterDetails { get; set; }
    }

    /// <summary>
    /// Model for an encounter detail.
    /// </summary>
    public class EncounterDetailEntry
    {
        /// <summary>
        /// The lowest level of the encounter detail.
        /// </summary>
        public int MinLevel { get; set; }

        /// <summary>
        /// The highest level of the encounter detail.
        /// </summary>
        public int MaxLevel { get; set; }

        /// <summary>
        /// The condition values that must be in effect for this encounter to occur.
        /// </summary>
        public List<EncounterConditionValueEntry> ConditionValues { get; set; }

        /// <summary>
        /// The percent chance that this encounter will occur.
        /// </summary>
        public int Chance { get; set; }
    }
}
